# Reading #2

- Due Tuesday, December 2, 2025, 11:59 pm.
- Turn in on canvas.

Paper in your own field
Due Tuesday, November 19, 2024, 10:00 am. Turn in on canvas.

Purpose: The purpose of this homework assignment is to help you start to think about your final exam project and explore some of the existing datsets in your field. If you would prefer to used one of the “suggested” datasets, they are listed below.

Read a Paper that will relate to your FINAL PROJECT
Identify and read a paper that includes an RNA-seq experiment.
State the paper you read
Briefly summarize the point of the paper (you don't need to read the whole paper, just get a gist of what the paper is about through the abstract and by skimming)
Briefly summarize the approach the researchers used in their RNA-seq analysis.
.You'll probably need to read the materials & methods section for this, or the STAR protocols, or a supplemental method
Report what software and versions were used?
Report Is this Paired end or single end?
Report whether the researchers include information on the data repository
:!: NOTE: The pipeline we are using works on paired-end sequencing. Please try to identify a dataset that uses paired-end sequencing. Please contact me if you want to use a single-end dataset.

:!: NOTE - microarray, ddPCR, minION, and pacBIO datasets will not work for this assignment.

HINTS

If you are going to use your OWN data for your final project and that data is not yet published,you can either find a publication in the same organism OR a similar type of experiment done in another field.
If you would like to do your Final Project on a published work relevant to your field, this is a good opportunity to discuss some ideas with your PI and search to see if there is a specific dataset that would be appealing.
If you don't have a lab, a project, or as yet don't know what you're planning to do, consider using the publication options below for this assignment.
OPTIONS

Here are some papers if you are a little bit stuck…

Yeast
Maksiutenko EM, Barbitoff YA, Danilov LG, Matveenko AG, Zemlyanko OM, Efremova EP, Moskalenko SE, Zhouravleva GA. (2024) Gene Expression Analysis of Yeast Strains with a Nonsense Mutation in the eRF3-Coding Gene Highlights Possible Mechanisms of Adaptation. Int J Mol Sci. Jun 7;25(12):6308. doi: 10.3390/ijms25126308.

Dong Y, Hu J, Fan L, and Chen Q. (2016) RNA-seq-based transcriptomic and metabolomic analysis reveal stress responses and programmed cell death induced by acetic acid in Saccharomyces cerevisiae. Scientific Reports. 7:4.

Arabidopsis
Herranz, Vandenbrink, Villacampa, Manzano, Poehlman, Feltus, Kiss, and Medina. (2019) RNAseq Analysis of the Response of Arabidopsis thaliana to Fractional Gravity Under Blue-Light Stimulation During Spaceflight. Front Plant Sci. 2019 Nov 26;10:1529.

Park et al., (2019) The transcriptome analysis of the Arabidopsis thaliana in response to the Vibrio vulnificus by RNA-sequencing. PLoS One. 2019 Dec 16;14(12):e0225976.

Mouse
Nardou R, Sawyer E, Song YJ, Wilkinson M, Padovan-Hernandez Y, de Deus JL, Wright N, Lama C, Faltin S, Goff LA, Stein-O'Brien GL, Dölen G. Psychedelics reopen the social reward learning critical period. (2023) Nature. Jun;618(7966):790-798. doi: 10.1038/s41586-023-06204-3.

Brady RA, Bruno VM, Burns DL. (2015) RNA-Seq Analysis of the Host Response to Staphylococcus aureus Skin and Soft Tissue Infection in a Mouse Model

Tuipulotu, Netzler, Lun, Mackenzie, and White1. (2017) RNA Sequencing of Murine Norovirus-Infected Cells Reveals Transcriptional Alteration of Genes Important to Viral Recognition and Antigen Presentation. Front. Immunol. 11 August
