# Course Introduction

Today's Objectives:
- Introduction to the course
- Understand the goal of RNA-seq
- Understand what we can learn from RNA-seq
- Become familiar with the “vanilla” RNA-seq protocol
- Learn how to log into ALPINE and install required software for this course

## Welcome to class!

- Prior LINUX experience - strongly encouraged
- Prior ALPINE experience - will be helpful but not required
- Prior R experience - will be helpful but not required

### Intro to instructor: 

- Jessica Hill, PhD

### Intro to class:

- Please read through the [Syllabus](../../Syllabus.md) to learn more about the class.
- This class is the last in a 3-part series. The other two classes are DSCI510 LINUX and DSCI511 Python. This class focuses on building pipelines and computational projects in LINUX and R using RNA-seq data analysis as an example.
- The course will be primarily accessed through this GitHub page. There is a Canvas page for the class and this will be the area to turn in assignments and see grades.

Please take this [brief survey](https://forms.gle/VuRskcNHvUHU5WeY6)

## Goals of the course

### By the end of this course, you should ...
- Have a grasp of the goals and process of RNA-seq experiments
- Understand RNA-seq experimental design
- Become familiar with common RNA-seq analysis steps: quality control, alignment, tabulation, differential expression, and clustering
- Be able to write or hack automated pipelines for RNA-seq analysis
- Be familiar with best practices

### What this course doesn't cover ...
- We will focus on Illumina sequencing. We don't cover Pac Bio or minION sequencing.
- We focus on RNA-seq on model organisms and humans, not on un-sequenced genomes
- We don't cover metagenomics
- We don't cover single-cell resolution sequencing, but we can talk about it
- However, many of the general principles and best practices will carry over, and everything you learn will be generalizable and relevant!

#### When I leave this course, will I be able to analyze my own sequencing data?

Probably, yes! You will definitely be able to design your experiments with deeper thought and insight. You'll be able to do a basic “vanilla” level analysis of your own RNA-seq data. More advanced projects will require you to take on new training and education on your own, but you will likely be able to study independently. You will also be able to have better conversations with computational collaborators whom you may need to bring in as your analysis becomes more advanced.

Continue on to [ALPINE](1_2_ALPINE.md)
